File: src/leetcode/python/easy/PalindromeLinkedList_234.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class PalindromeLinkedList_234(object):
    def isPalindrome(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        if not head or not head.next:
            return True
        fast = head
        mid = head
        while fast and fast.next:
            fast, mid = fast.next.next, mid.next

        pt = self.reverseList(mid)
        while head and pt:
            if head.val != pt.val:
                return False
            head, pt = head.next, pt.next
        return True

    def reverseList(self, head):
        if not head:
            return None
        elif not head.next:
            return head
        tail, head.next = head.next, None
        while tail:
            tmp = tail.next
            tail.next, head, tail = head, tail, tmp
        return head

File: src/leetcode/java/PalindromeLinkedList_234.java
package leetcode.java;

public class PalindromeLinkedList_234 {
  public boolean isPalindrome(ListNode head) {
    if (head == null || head.next == null) return true;
    // 1. find the middle pointer
    ListNode mid = head;
    ListNode fast = head;
    while (fast != null && fast.next != null) {
      fast = fast.next.next;
      mid = mid.next;
    }
    // 2. reverse the second half
    ListNode half = reverseList(mid);
    // 3. compare two halves
    while (head != null && half != null) {
      if (head.val != half.val) return false;
      head = head.next;
      half = half.next;
    }
    return true;
  }

  private static ListNode reverseList(ListNode head) {
    if (head == null) return null;
    else if (head.next == null) return head;

    ListNode tmp = null;
    ListNode next = head.next;
    head.next = null;
    while (next != null) {
      tmp = next.next;
      next.next = head;
      head = next;
      next = tmp;
    }
    return head;
  }

  class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
      val = x;
      next = null;
    }
  }
}
